### DISCUSSION TITLE
Support for Ein Clone

### DISCUSSION THREAD

**Discussion by Limits0x** [GitoxideLabs/gitoxide#300] [Q&A] [answered]
As written in the `README`, Ein current does not have any support for cloning/fetching repos.
I would like to contribute to this project and would like instructions on how I should go about implementing clone and fetch in Ein.

**Comment by Byron** [comment, marked as answer]
I have converted the issue into into a question, which may very well serve to spawn more concrete issues later.

For rough steps on what's needed to finish a clone [this checklist](https://github.com/Byron/gitoxide/blob/0ee70866dcec9877b09e0ec77df303396145b1f5/tasks.md#L3) should be useful.

In theory, the last big missing piece to support clones is to produce an index from a tree, as well as the ability to checkout a worktree from that index.

In order to estimate where you could help, please let me ask…

* …how much time you have per week and…
* …what's your level of experience with Rust?

**Reply by Limits0x** [reply]
> …how much time you have per week and…

About ~12 hours...
> …what's your level of experience with Rust?

I have been writing rust code for about ~2 years now. (Looking for OSS projects to contribute to :))
> For rough steps on what's needed to finish a clone this checklist should be useful.

Thanks for linking this! I'll look into it.

**Reply by Byron** [reply]
Great, thanks!

You could probably start by getting acquainted with `git-index` and see how all pieces come together. Before getting started, lets synchronize here (for now) as I might soon get started with additional work on `git-index`.

**Reply by Limits0x** [reply]
> In theory, the last big missing piece to support clones is to produce an index from a tree, as well as the ability to checkout a worktree from that index.

What is a `tree` and an `index` ? Is there a document where these terms are defined ?

**Reply by Limits0x** [reply]
By `tree`, are you referencing [Tree Objects](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects) ?

**Reply by Byron** [reply]
The latter is correct. Crates of interest would be `git-object` and `git-traverse`. 

> What is a tree and an index ? Is there a document where these terms are defined ?

That's probably the documentation you referenced.

**Reply by Limits0x** [reply]
I suppose [this](https://github.com/Byron/gitoxide/blob/d726eb5b96479881675a05f98e2003d3a7aeef23/git-object/src/lib.rs#L220) is the `tree` and [this](https://github.com/Byron/gitoxide/blob/d726eb5b96479881675a05f98e2003d3a7aeef23/git-index/src/lib.rs#L63) is the `index`.
So, right now there is no implementation to convert a `tree` into an `index` ?

**Reply by Byron** [reply]
Yes, that's correct. Except that it's better to take `TreeRef` as input instead of an owned version of a tree.
There is no implementation to obtain an index from a tree.

As I plan to work on this tomorrow, it might make more sense to see how a worktree can be checked out from an index. `git-worktree` is entirely empty and it's unclear what exactly this is for `git` (as I didn't look into it yet and there certainly are plenty of complexities). This might be an interesting area for investigation.

**Reply by Limits0x** [reply]
> Yes, that's correct. Except that it's better to take `TreeRef` as input instead of an owned version of a tree.
> There is no implementation to obtain an index from a tree.
> 
> As I plan to work on this tomorrow, it might make more sense to see how a worktree can be checked out from an index. `git-worktree` is entirely empty and it's unclear what exactly this is for `git` (as I didn't look into it yet and there certainly are plenty of complexities). This might be an interesting area for investigation.

Sure. I'll look into `git-worktree`.

**Reply by Byron** [reply]
In theory, it would be possible to checkout from a tree directly, too, without the need for an intermediate index (which is generated from the same tree after all). Both would work. Note that the implementation should ideally be optionally multi-threaded, which is probably easiest if one is going through an index. It's valuable nonetheless to have a version of checkout that is single-threaded and uses a tree as source directly.

In the end, I recommend doing whatever seems easiest, without sacrificing correctness. There are a few special cases around file modes, newline styles, file filters and and case-insensitive file systems.

**Reply by Limits0x** [reply]
Do you want to convert a `tree`/`index` to a `worktree` in memory or write it to disk ?

**Reply by Limits0x** [reply]
```bash
$ git worktree add -b new-working-branch ../sandbox-worktree-dir master
$ tree .git/worktrees
.git/worktrees
└── sandbox-worktree-dir
    ├── commondir
    ├── gitdir
    ├── HEAD
    ├── index
    ├── logs
    │   └── HEAD
    └── ORIG_HEAD
```
The `commondir` file will always have `../..`. It is [hard-coded](https://github.com/git/git/blob/297ca895a27a6bbdb7906371d533f72a12ad25b2/builtin/worktree.c#L336) in the git source code.
The `gitdir` file will contain the path of the `.git` dir. How do I get this in the program ?
The `HEAD` file contains `ref: refs/heads/sandbox-worktree-dir`. I suppose I need to create this `ref` ? How do I do this ?
The `ORIG_HEAD` file contains the last commit. Does `gix` have concepts of commits yet ?

```bash
$ tree -a ../sandbox-worktree-dir
../sandbox-worktree-dir
├── .git
```
^ This is the newly created `worktree` which has a `.git` file.
The `.git` file contains the absolute path to `.git/worktrees/sandbox-worktree-dir`: 
`gitdir: /home/.../sandbox/.git/worktrees/sandbox-worktree-dir`

**Reply by Byron** [reply]
It's probably very unintuitive that `git-worktree` has only a very 'general' relationship with the `git worktree` command, as its meant to be a plumbing crate. This means it can focus entirely on what it would mean to checkout/create directories and files from an index.

For that, it would probably need nothing more than `git_index::State` and the root directory of the files to be created. Since it's the first clone, it may reasonably expect this directory to be empty even, so can forgo a lot of checks it would have to do otherwise (I presume, didn't look into it).

That could be a starting point from which one can expand later, as more abstractions emerge along with new challenges, like implementing `git worktree` like functionality, or implementing something like `git checkout`.

I hope that helps.

> The HEAD file contains ref: refs/heads/sandbox-worktree-dir. I suppose I need to create this ref ? How do I do this ?

This can be left to the enclosing code, which would probably live in `gitoxide-core` initially, which is used to put all the pieces together and get a first 'clone' like command going.

**Reply by Limits0x** [reply]
Thanks for the clarification! I thought `git-worktree` was the same as `git worktree`.

**Reply by Byron** [reply]
Indeed, `git-worktree` is a place that handles (or creates) worktree checkouts and probably all interactions that alter the `git-index` state which is like the in-memory version of a worktree checkout.

**Reply by Byron** [reply]
By the way, I do keep track of findings relevant to checking out a worktree [in this tracking issue](https://github.com/Byron/gitoxide/issues/301) . It's rough but maybe helpful nonetheless.

**Reply by Limits0x** [reply]
I may have a working implementation to produce a worktree from an index:
It extracts the paths of all the Entries in the index and then recursively copies them to the target directory. Is this correct ?
Also, if it is correct, how do I test it (run) ?

**Reply by Byron** [reply]
It sounds correct, but for further guidance it would be best to create a draft PR on which I can comment. Thank you.

**Reply by Limits0x** [reply]
[Draft PR](https://github.com/Byron/gitoxide/pull/315)
